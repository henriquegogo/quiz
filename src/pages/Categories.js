import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { categories } from '../actions/categories';

class Categories extends Component {
  componentDidMount() {
    this.props.dispatch(categories());
  }

  render() {
    const { categories } = this.props;

    return (
      <Fragment>
        <h1>Categorias</h1>
        {categories.length === 0 && 'Carregando...'}
        <div className='card-group'>
          {categories.map(category =>
            <Link className='card' key={category.id} to={'/categories/' + category.id}>{category.name}</Link>
          )}
        </div>
      </Fragment>
    );
  }
}

export default connect(({ categories }) => categories)(Categories);
