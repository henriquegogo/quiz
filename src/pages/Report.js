import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import './Report.scss';

class Report extends Component {

  state = {
    total_hits: 0,
    total_mistakes: 0,
    easy_hits: 0,
    easy_mistakes: 0,
    medium_hits: 0,
    medium_mistakes: 0,
    hard_hits: 0,
    hard_mistakes: 0
  }
  
  componentDidMount() {
    let answers = [{"category":"General Knowledge","type":"multiple","difficulty":"medium","question":"When was the Declaration of Independence approved by the Second Continental Congress?","correct_answer":"July 2, 1776","incorrect_answers":["May 4, 1776","June 4, 1776","July 4, 1776"],"options":["July 4, 1776","June 4, 1776","July 2, 1776","May 4, 1776"],"selected_answer":"May 4, 1776","correct":true,"date_time":"2019-09-01T19:54:06.749Z"},{"category":"General Knowledge","type":"multiple","difficulty":"medium","question":"What name represents the letter &quot;M&quot; in the NATO phonetic alphabet?","correct_answer":"Mike","incorrect_answers":["Matthew","Mark","Max"],"options":["Mike","Max","Mark","Matthew"],"selected_answer":"Matthew","correct":false,"date_time":"2019-09-01T19:54:08.721Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"What is &quot;dabbing&quot;?","correct_answer":"A dance","incorrect_answers":["A medical procedure","A sport","A language"],"options":["A dance","A language","A sport","A medical procedure"],"selected_answer":"A medical procedure","correct":false,"date_time":"2019-09-01T19:54:09.873Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"What is the name of the Jewish New Year?","correct_answer":"Rosh Hashanah","incorrect_answers":["Elul","New Year","Succoss"],"options":["Succoss","New Year","Rosh Hashanah","Elul"],"selected_answer":"Elul","correct":false,"date_time":"2019-09-01T19:54:10.775Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"What was the nickname given to the Hughes H-4 Hercules, a heavy transport flying boat which achieved flight in 1947?","correct_answer":"Spruce Goose","incorrect_answers":["Noah&#039;s Ark","Fat Man","Trojan Horse"],"options":["Spruce Goose","Fat Man","Noah&#039;s Ark","Trojan Horse"],"selected_answer":"Trojan Horse","correct":false,"date_time":"2019-09-01T19:54:11.633Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"The New York Times slogan is, &ldquo;All the News That&rsquo;s Fit to&hellip;&rdquo;","correct_answer":"Print","incorrect_answers":["Digest","Look","Read"],"options":["Look","Print","Read","Digest"],"selected_answer":"Digest","correct":false,"date_time":"2019-09-01T19:54:12.601Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"What do the letters of the fast food chain KFC stand for?","correct_answer":"Kentucky Fried Chicken","incorrect_answers":["Kentucky Fresh Cheese","Kibbled Freaky Cow","Kiwi Food Cut"],"options":["Kentucky Fried Chicken","Kiwi Food Cut","Kibbled Freaky Cow","Kentucky Fresh Cheese"],"selected_answer":"Kentucky Fresh Cheese","correct":false,"date_time":"2019-09-01T19:54:13.465Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"Which candy is NOT made by Mars?","correct_answer":"Almond Joy","incorrect_answers":["M&amp;M&#039;s","Twix","Snickers"],"options":["Almond Joy","Snickers","Twix","M&amp;M&#039;s"],"selected_answer":"M&amp;M&#039;s","correct":false,"date_time":"2019-09-01T19:54:14.473Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"Which of the following is the IATA code for Manchester Airport?","correct_answer":"MAN","incorrect_answers":["EGLL","LHR","EGCC"],"options":["EGCC","LHR","MAN","EGLL"],"selected_answer":"EGLL","correct":false,"date_time":"2019-09-01T19:54:15.697Z"},{"category":"General Knowledge","type":"multiple","difficulty":"easy","question":"According to the nursery rhyme, what fruit did Little Jack Horner pull out of his Christmas pie?","correct_answer":"Plum","incorrect_answers":["Apple","Peach","Pear"],"options":["Apple","Plum","Pear","Peach"],"selected_answer":"Peach","correct":true,"date_time":"2019-09-01T19:54:16.801Z"}];

    const new_state = answers.reduce((result, answer) => {
      answer.correct ? result.total_hits++ : result.total_mistakes++;
      return result;
    }, { ...this.state });

    this.setState(new_state);
  }


  render() {
    const {
      total_hits,  total_mistakes,
      easy_hits,   easy_mistakes,
      medium_hits, medium_mistakes,
      hard_hits,   hard_mistakes
    } = this.state;

    return (
      <div className='report'>
        <div className='report-head'>
          <img src='./mascot.png' alt='Mascot' />
          <div className='report-congrats'>
            <h1>Congrats!</h1>
            <h2>You finished the test</h2>
          </div>
        </div>
        <div className='report-body'>
          <h3>See your performance</h3>
          <div className='report-resume'>
            <div className='report-resume-points'>
              <div>{total_hits}</div>
              <span>hits</span>
            </div>
            <div className='report-resume-points'>
              <div>{total_mistakes}</div>
              <span>mistakes</span>
            </div>
          </div>
          <div className='report-detail'>
            <div>
              <b>Easy</b><br />
              Hits: {easy_hits}<br />
              Mistakes: {easy_mistakes}
            </div>
            <div>
              <b>Medium</b><br />
              Hits: {medium_hits}<br />
              Mistakes: {medium_mistakes}
            </div>
            <div>
              <b>Hard</b><br />
              Hits: {hard_hits}<br />
              Mistakes: {hard_mistakes}
            </div>
          </div>
          <Link to='/'>
            <button>Back to start</button>
          </Link>
        </div>
      </div>
    );
  }
}

export default connect(({ questions }) => questions)(Report);
