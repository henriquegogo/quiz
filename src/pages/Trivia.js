import React, { Component, Fragment, createRef } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { questions, clearQuestions, answerQuestion } from '../actions/questions';
import Level from '../components/Level'
import Modal from '../components/Modal'

const MAX_QUESTIONS = 10;

class Trivia extends Component {

  state = {
    selected_answer: null,
    question_index: 0,
    correct: null
  }

  selectAnswer = (e) => {
    this.setState({ selected_answer: e.target.value });
  }

  submit = (e, question) => {
    e.preventDefault();
    const correct = this.state.selected_answer === question.correct_answer;
    this.setState({ correct });

    const answer = {
      ...question,
      selected_answer: this.state.selected_answer,
      correct,
      date_time: new Date()
    };
    this.props.dispatch(answerQuestion(answer));
  }

  close = (e) => {
    e.preventDefault();
    this.props.history.push('/');
  }

  next = (e) => {
    e.preventDefault();
    const { question_index } = this.state;
    const { answers } = this.props;

    if (answers.length < MAX_QUESTIONS) {
      this.form.current.reset();
      this.setState({
        selected_answer: null,
        question_index: question_index + 1,
        correct: null
      });
    }
    else {
      console.log('The end');
    }
  }

  componentDidMount() {
    this.props.dispatch(questions(this.props.category));
  }

  componentWillUnmount() {
    this.props.dispatch(clearQuestions());
  }

  render() {
    const { selected_answer, question_index, correct } = this.state;
    const question = this.props.questions[question_index] || {};

    return !question.question ? 'Carregando...' : (
      <Fragment>
        <a href='#close' onClick={this.close}><i>ⓧ</i> Fechar</a>
        <h2>{question.category}</h2>
        <section>
          <Level difficulty={question.difficulty} />
          <h3>Questão {question_index + 1}</h3>
          {/* Needs to use dangerouslySetInnerHTML because api returns some HTML encoded text */}
          <p dangerouslySetInnerHTML={{ __html: question.question }} />
          <form className='answers' ref={this.form = createRef()}
            onChange={this.selectAnswer}
            onSubmit={e => this.submit(e, question)}>
            {question.options && question.options.map(option =>
              <label key={btoa(option)}>
                <input type='radio' name='option' value={option} />
                <span dangerouslySetInnerHTML={{ __html: option }} />
              </label>
            )}
            <button type='submit' disabled={!selected_answer}>Responder</button>
            {correct !== null && <Modal correct={correct} next={this.next} />}
          </form>
        </section>
      </Fragment>
    );
  }
}

export default connect(({ questions }) => questions)(withRouter(Trivia));
