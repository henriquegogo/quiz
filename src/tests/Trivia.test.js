import React from 'react';
import { shallow } from 'enzyme';
import { Trivia } from '../pages/Trivia';

describe('Trivia', () => {
  let component, dispatch, questions, answers;

  beforeEach(() => {
    questions = [
      { "category": "General Knowledge", "type": "multiple", "difficulty": "medium",
        "question": "Which language is NOT Indo-European?",
        "correct_answer": "Hungarian", "incorrect_answers": ["Russian","Greek","Latvian"],
        "options": ["Hungarian","Russian","Greek","Latvian"]
      },
      { "category": "General Knowledge", "type": "multiple", "difficulty": "medium",
        "question": "What character was once considered to be the 27th letter of the alphabet?",
        "correct_answer": "Ampersand", "incorrect_answers":["Interrobang","Tilde","Pilcrow"],
        "options": ["Ampersand","Interrobang","Tilde","Pilcrow"]
      }
    ];
    answers = [];
    dispatch = jest.fn();
    component = shallow(
      <Trivia questions={questions} answers={answers} dispatch={dispatch} />
    );
  });

  it('renderize properly', () => {
    expect(component).toMatchSnapshot();
  });

  it('requests questions from api', () => {
    expect(dispatch).toHaveBeenCalled();
  });

  it('shows a question with medium difficulty', () => {
    expect(component.find('Level').html().includes('Medium')).toBeTruthy();
  });

  it('shows multiple choices', () => {
    expect(component.find('.answers input[name="option"]').length).toBe(4);
  });

  it('has a "Send answer" button', () => {
    expect(component.find('button[type="submit"]').text()).toBe('Send answer');
  });
});
