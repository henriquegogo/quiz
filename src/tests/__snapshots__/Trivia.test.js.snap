// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trivia renderize properly 1`] = `ShallowWrapper {}`;
